skimage/transform/_distance_transform.pyx

```diff
@@ -0,0 +1,143 @@
+import numpy as np
+import warnings
+cimport numpy as np
+
+cdef f(bint p):
+    cdef double out
+    if p == 0:
+        out = 0
+    out = np.finfo(np.float64).max
+    print(warnings.warn(p,out))
+    return out
+
+cdef euclidean_dist(double a, double b, double c):
+    cdef double out = (a-b)**2+c
+    return out
+
+cdef euclidean_meet(double a, double b, np.ndarray[double, ndim=1] f):
+    cdef double out = (f[a]+a**2-f[b]-b**2)/(2*a-2*b)
+    return out
+
+cdef manhattan_dist(double a, double b, double c):
+    cdef double out = (a-b)**2+c
+    return out
+
+cdef manhattan_meet(double a, double b, np.ndarray[double, ndim=1] f):
+    cdef double s
+    s = (a + f[a] + b - f[b]) / 2
+    if manhattan_dist(a,s,f[a])==manhattan_dist(b,s,f[b]):
+        return s
+    s = (a - f[a] + b + f[b]) / 2
+    if manhattan_dist(a,s,f[a])==manhattan_dist(b,s,f[b]):
+        return s
+    if manhattan_dist(a,a,f[a]) > manhattan_dist(b,a,f[b]):
+        return <double>np.finfo(np.float64).max
+    return <double>np.finfo(np.float64).min
+
+def _generalized_distance_transform_1d_euclidean(double[:] arr, double[:] cost_arr,
+                                       bint isfirst, double[::1] domains,
+                                       Py_ssize_t[::1] centers, double[::1] out):
+    cdef Py_ssize_t length = len(arr)
+    cdef Py_ssize_t i, rightmost, current_domain
+    cdef double intersection
+
+    if isfirst:
+        for i in range(length):
+            cost_arr[i] = f(<bint>arr[i])
+
+    rightmost = 0
+    domains[0] = -np.inf
+    domains[1] = np.inf
+    
+    for i in range(1,length):
```

Comment:
I would only start to release the gil once you have the algorithm down, but generally, the strategy is:
```
    with nogil:
        for i in range(1, length):
```

Reply:
Cython knows about the memory layout of numpy arrays. Therefore the python operator `[ ]` can be avoided and it will use standard `c-style` indexing. 

https://cython.readthedocs.io/en/latest/src/userguide/numpy_tutorial.html#numpy-tutorial

Try to get to the memoryviews part, and efficient indexing. Have a look at some of the cython code in scikit-image, it uses many of these idioms. I can help once your code is more ready to be optimized. We can get some benchmarks going to to make sure that we are improving

Reply:
Just ping me when you are ready for a performance oriented review, and we can get to that bridge when we need to.